action/evm.go

```diff
@@ -0,0 +1,231 @@
+package action
+
+import (
+	"fmt"
+	"hash/fnv"
+	"math/big"
+
+	"github.com/Oneledger/protocol/data/keys"
+	"github.com/ethereum/go-ethereum/common"
+	ethcmn "github.com/ethereum/go-ethereum/common"
+	ethcore "github.com/ethereum/go-ethereum/core"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	ethvm "github.com/ethereum/go-ethereum/core/vm"
+	ethparams "github.com/ethereum/go-ethereum/params"
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+	"github.com/tendermint/tendermint/version"
+)
+
+// AbciHeaderToTendermint is a util function to parse a tendermint ABCI Header to
+// tendermint types Header.
+func AbciHeaderToTendermint(header *abci.Header) tmtypes.Header {
+	return tmtypes.Header{
+		Version: version.Consensus{
+			Block: version.Protocol(header.Version.Block),
+			App:   version.Protocol(header.Version.App),
+		},
+		ChainID: header.ChainID,
+		Height:  header.Height,
+		Time:    header.Time,
+
+		LastBlockID: tmtypes.BlockID{
+			Hash: header.LastBlockId.Hash,
+			PartsHeader: tmtypes.PartSetHeader{
+				Total: int(header.LastBlockId.PartsHeader.Total),
+				Hash:  header.LastBlockId.PartsHeader.Hash,
+			},
+		},
+		LastCommitHash: header.LastCommitHash,
+		DataHash:       header.DataHash,
+
+		ValidatorsHash:     header.ValidatorsHash,
+		NextValidatorsHash: header.NextValidatorsHash,
+		ConsensusHash:      header.ConsensusHash,
+		AppHash:            header.AppHash,
+		LastResultsHash:    header.LastResultsHash,
+
+		EvidenceHash:    header.EvidenceHash,
+		ProposerAddress: header.ProposerAddress,
+	}
+}
+
+// // HashFromContext returns the Ethereum Header hash from the context's protocol
+// // block header.
+// func HashFromHeader(header *abci.Header) ethcmn.Hash {
+// 	// cast the ABCI header to tendermint Header type
+// 	tmHeader := AbciHeaderToTendermint(header)
+
+// 	// get the Tendermint block hash from the current header
+// 	tmBlockHash := tmHeader.Hash()
+
+// 	// NOTE: if the validator set hash is missing the hash will be returned as nil,
+// 	// so we need to check for this case to prevent a panic when calling Bytes()
+// 	if tmBlockHash == nil {
+// 		return ethcmn.Hash{}
+// 	}
+
+// 	return ethcmn.BytesToHash(tmBlockHash.Bytes())
+// }
+
+// GetHashFn implements vm.GetHashFunc for protocol. It handles 3 cases:
```

Comment:
3 cases? but you just show 2.

Reply:
Not removed a comment from the etheremint. Will remove as we do not need to make a same logic as there.